modules/contracts/apps/subscription/src/handlers/execute.rs

```diff
@@ -0,0 +1,351 @@
+            .amount
+            .checked_div_floor(config.subscription_cost_per_week)?
+            .u128() as u64;
+        paid_for_weeks * 7
+    };
+    if let Some(mut active_sub) = maybe_subscriber {
+        // Subscriber is active, update balance
+        active_sub.expiration_timestamp = active_sub.expiration_timestamp.plus_days(paid_for_days);
+        SUBSCRIBERS.save(deps.storage, &subscriber_addr, &active_sub)?;
+    } else {
+        // Subscriber is (re)activating his subscription.
+        if asset.amount.u128() < required_payment.u128() {
+            return Err(SubscriptionError::InsufficientPayment(
+                required_payment.u128() as u64,
+                deposit_info.to_string(),
+            ));
+        }
+        let maybe_old_client = EXPIRED_SUBSCRIBERS.may_load(deps.storage, &subscriber_addr)?;
+        // if old client
+        if let Some(mut old_client) = maybe_old_client {
+            EXPIRED_SUBSCRIBERS.remove(deps.storage, &subscriber_addr);
+            old_client.expiration_timestamp = env.block.time.plus_days(paid_for_days);
+            old_client.last_emission_claim_timestamp = env.block.time;
+            SUBSCRIBERS.save(deps.storage, &subscriber_addr, &old_client)?;
+            return Ok(Response::new().add_attributes(attrs).add_message(
+                // Send the received asset to the proxy
+                asset.transfer_msg(base_state.proxy_address)?,
+            ));
+        } else {
+            // New client
+            let new_sub = Subscriber {
+                expiration_timestamp: env.block.time.plus_days(paid_for_days),
+                last_emission_claim_timestamp: env.block.time,
+            };
+            SUBSCRIBERS.save(deps.storage, &subscriber_addr, &new_sub)?;
+        }
+        let mut subscription_state = SUBSCRIPTION_STATE.load(deps.storage)?;
+        INCOME_TWA.accumulate(
+            &env,
+            deps.storage,
+            Decimal::from_atomics(Uint128::from(subscription_state.active_subs), 0)?
+                * config.subscription_cost_per_week,
+        )?;
+        subscription_state.active_subs += 1;
+        SUBSCRIPTION_STATE.save(deps.storage, &subscription_state)?;
+    }
+
+    Ok(app.tag_response(
+        Response::new().add_attributes(attrs).add_message(
+            // Send the received asset to the proxy
+            asset.transfer_msg(base_state.proxy_address)?,
+        ),
+        "pay",
+    ))
+}
+
+pub fn unsubscribe(
+    deps: DepsMut,
+    env: Env,
+    app: SubscriptionApp,
+    unsubscribe_addrs: Vec<String>,
+) -> SubscriptionResult {
+    let unsubscribe_addrs: Vec<Addr> = unsubscribe_addrs
+        .iter()
+        .map(|human| deps.api.addr_validate(human))
+        .collect::<StdResult<_>>()?;
+    let mut subscription_state = SUBSCRIPTION_STATE.load(deps.storage)?;
+    let subscription_config = SUBSCRIPTION_CONFIG.load(deps.storage)?;
+    let mut unsubscribed_addrs: Vec<String> = vec![];
```

Comment:
Hm, perhaps there is a better way to name the `unsubscribe_addrs` and `unsubscribed_addrs` variables, they are kind of confusing.

Reply:
`canceled_subs` for addrs that are indeed "canceled" is better?